### PR TITLE
CT 1894 log partial parsing var changes and sort cli vars before hashing

### LINKED ISSUE #6710: [CT-1894] Add logging to improve diagnosing of unable to partial parsing because vars have changed
It's difficult to diagnose situations in which partial parsing is aborted due to changes in vars. Add logging to make it easier to determine what changed.

### PR DESCRIPTION
**Pull request by gshank**
resolves #6710

### Description

If the order of vars passed on the command line changes, then partial parsing will be aborted. Sort the vars before hashing in order to avoid unnecessary reparsing.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: core/dbt/events/types.py [hunk 1]
```diff
@@ -843,6 +843,15 @@ def message(self) -> str:
         return f"Unable to do partial parsing because {self.reason}"
 
 
+@dataclass
+class StateCheckVarsHash(DebugLevel, pt.StateCheckVarsHash):
+    def code(self):
+        return "I025"
+
+    def message(self) -> str:
+        return "checksum: {self.checksum}, vars: {self.vars}, profile: {self.profile}, target: {self.target}, version: {self.version}"
```
Comment (dbeatty10) [inline core/dbt/events/types.py:852]: Is this supposed to be an f-string?
```suggestion
        return f"checksum: {self.checksum}, vars: {self.vars}, profile: {self.profile}, target: {self.target}, version: {self.version}"
```

File: core/dbt/parser/manifest.py [hunk 2]
```diff
@@ -702,16 +711,28 @@ def build_manifest_state_check(self):
...
         # to not pass, it doesn't matter.  If we move to more granular checking
         # of env_vars, that would need to change.
+        # We are using the parsed cli_vars instead of config.args.vars, in order
+        # to sort them and avoid reparsing because of ordering issues.
+        stringified_cli_vars = pprint.pformat(config.cli_vars)
         vars_hash = FileHash.from_contents(
             "\x00".join(
                 [
-                    getattr(config.args, "vars", "{}") or "{}",
+                    stringified_cli_vars,
```
Comment (dbeatty10) [inline core/dbt/parser/manifest.py:720]: Trying to give a close look at the old code versus the new code through the lens of backwards compatibility.

Can `config.cli_vars` ever be `None` or any empty string (`""`)? Or is it always a dictionary, even if an empty dictionary (`{}`)?

Just trying to convince myself that `pprint.pformat(config.cli_vars)` will be functionally equivalent to `getattr(config.args, "vars", "{}") or "{}"` in the degenerate case where `config.cli_vars` has "nothing".

Comment (gshank) [reply core/dbt/parser/manifest.py:720]: cli_vars is defined as a Dict[str, Any]. It's not an optional Dict, so it should never be None. In addition, parse_cli_vars will throw an error if it's not parsable to a dictionary.

**Review by dbeatty10 — APPROVED** [review]
Looks good to me 👍

**Review by peterallenwebb — APPROVED** [review]

**Comment by jtcohen6** [comment]
@gshank Is this something you'd feel comfortable backporting for inclusion in a v1.4.x patch release? The changes look quite minor:
- new logging
- functional change to sort & stringify `vars` before hashing

(cc @bossnunta)

**Comment by gshank** [comment]
Yes, should be pretty straightforward to backport. Will do that.

**Pull request merged**